lib/metadata/chordMetadata.hpp

```diff
@@ -392,6 +399,128 @@ class chordMetadata : public metadataObject {
+        return beamCoord(metadata);
+    }
+
+    // TODO: add set_beam_coord
+
+    int64_t get_fpga_seq_num() const {
+        return metadata[jsonMetadata::FPGA_SEQ_NUM].template get<int64_t>();
+    }
+
+    void set_fpga_seq_num(const int64_t fpga_seq_num) {
+        metadata[jsonMetadata::FPGA_SEQ_NUM] = fpga_seq_num;
+    }
+
+    int get_nfreq() const {
+        return this->nfreq;
+    }
+
+    // TODO: this should really be a freq_id_t array
+    const int* get_coarse_freq() const {
+        return this->coarse_freq;
+    }
+
+    void set_coarse_freq(const int nfreq, const int* coarse_freq) {
+        assert(nfreq < CHORD_META_MAX_FREQ);
+        this->nfreq = nfreq;
+        for(int i = 0 ; i < nfreq ; ++i) {
+            this->coarse_freq[i] = *coarse_freq++;
+        }
+    }
+
+    // TODO: remove this, its redundant
+    struct timespec get_gps_time() const {
+        const Telescope& tel = Telescope::instance();
+        return tel.to_time(this->get_fpga_seq_num());
+    }
+
+    // TODO: remove this, it's not setting anything anymore (and assumes that
+    // fpga_seq_num is set)
+    void set_gps_time(const timespec gps_time) {
+        const Telescope& tel = Telescope::instance();
+        const timespec my_gps_time = tel.to_time(this->get_fpga_seq_num());
+        assert(gps_time.tv_sec == my_gps_time.tv_sec);
+        assert(gps_time.tv_nsec == my_gps_time.tv_nsec);
+    }
+
+    /// The number of bad inputs in the RFI systems bad input list.
+    /// This value is mostly needed for renormalization of the SK values.
+    uint32_t get_rfi_num_bad_inputs() const {
+        return metadata[jsonMetadata::RFI_NUM_BAD_INPUTS].template get<uint32_t>();
+    }
+
+    void set_rfi_num_bad_inputs(const uint32_t rfi_num_bad_inputs) {
+        metadata[jsonMetadata::RFI_NUM_BAD_INPUTS] = rfi_num_bad_inputs;
+    }
+
+    /// The number of FPGA frames flagged as containing RFI.
+    /// NOTE: This value might contain overlap with lost samples, so it can count
+    /// missing samples as samples with RFI.  For renormalization this value
+    /// should NOT be used, use @c lost_timesamples instead.
+    /// This value will be filled even if RFI zeroing is disabled.
+    int32_t get_rfi_flagged_samples() const {
+        return metadata[jsonMetadata::RFI_FLAGGED_SAMPLES].template get<int32_t>();
+    }
+
+    void set_rfi_flagged_samples(const int32_t flagged_samples)  {
+        // very much non-atomic, due to json dict entry creation
+        metadata[jsonMetadata::RFI_FLAGGED_SAMPLES] = flagged_samples;
+    }
+
+    int32_t get_lost_timesamples() const {
+        return metadata[jsonMetadata::LOST_TIMESAMPLES].template get<int32_t>();
+    }
+
+    void set_lost_timesamples(int32_t lost_timesamples) {
+        // very much non-atomic, due to json dict entry creation
+        metadata[jsonMetadata::LOST_TIMESAMPLES] = lost_timesamples;
+    }
+
+    void atomic_add_lost_timesamples(const int32_t lost_samples) {
```

Comment:
We need to rethink metadata changes where more than one producer could be setting values.   This is very much an edge case, but still needs to be accounted for.   My vote would be to add a mutex to the metadata object and then we can do this safely.

Reply:
lost samples is currently one such where there are multiple threads accumulating onto the same metadata :-(